cli/cmd/healthcheck_extensions.go

```diff
@@ -0,0 +1,315 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/briandowns/spinner"
+	"github.com/linkerd/linkerd2/pkg/healthcheck"
+	"github.com/linkerd/linkerd2/pkg/version"
+	"github.com/mattn/go-isatty"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// glob is satisfied by filepath.Glob.
+type glob func(string) ([]string, error)
+
+// extension contains the full path of an extension executable. If it's a
+// a built-in extension, the builtin field is populated with its name.
+type extension struct {
+	path    string
+	builtin string
+}
+
+var (
+	builtInChecks = map[string]struct{}{
+		"jaeger":       {},
+		"multicluster": {},
+		"viz":          {},
+	}
+)
+
+// findExtensions searches the path for all linkerd-* executables and returns a
+// slice of check commands, and a slice of missing checks.
+func findExtensions(pathEnv string, glob glob, exec utilsexec.Interface, nsLabels []string) ([]extension, []string) {
+	cliExtensions := findCLIExtensionsOnPath(pathEnv, glob, exec)
+
+	// first, collect config extensions that are "always" enabled
+	extensions, checksSeen := findAlwaysChecks(cliExtensions, exec)
+
+	labelMap := map[string]struct{}{}
+	for _, label := range nsLabels {
+		if _, ok := checksSeen[label]; !ok {
+			labelMap[label] = struct{}{}
+		}
+	}
+
+	// second, collect on-cluster extensions
+	for _, e := range cliExtensions {
+		suffix := suffix(e)
+		if _, ok := labelMap[suffix]; ok {
+			extensions = append(extensions, extension{path: e})
+			delete(labelMap, suffix)
+		}
+	}
+
+	// third, collect built-in extensions
+	for label := range labelMap {
+		if _, ok := builtInChecks[label]; ok {
+			extensions = append(extensions, extension{path: os.Args[0], builtin: label})
+			delete(labelMap, label)
+		}
+	}
+
+	// anything left in labelMap is a missing executable
+	missing := []string{}
+	for label := range labelMap {
+		missing = append(missing, fmt.Sprintf("linkerd-%s", label))
+	}
+
+	sort.Slice(extensions, func(i, j int) bool {
+		if extensions[i].path != extensions[j].path {
+			_, filename1 := filepath.Split(extensions[i].path)
+			_, filename2 := filepath.Split(extensions[j].path)
+			return filename1 < filename2
+		}
+		return extensions[i].builtin < extensions[j].builtin
+	})
+	sort.Strings(missing)
+
+	return extensions, missing
+}
+
+// findCLIExtensionsOnPath searches the path for all linkerd-* executables and
+// returns a slice of unique filepaths.
```

Comment:
let's clarify what we mean by unique here: if multiple executables have the same name, only the one which comes earliest in the pathEnv is returned.